app/views/viewlets/change-version-jsx.jsx

```diff
@@ -0,0 +1,84 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2012-2013 Canonical Ltd.
+
+This program is free software: you can redistribute it and/or modify it under
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+
+YUI.add('change-version-view-jsx', function(Y) {
+
+  var ns = Y.namespace('juju.viewlets');
+
+  ns.ChangeVersionList = React.createClass({
+    // React is not MVC use utility classes and pass them into views.
+    // If it doesn't involve the UI do not put it in the component.
+    _upgradeService: function(upgradeTo) {
+      var props = this.props,
+          env = props.viewletManager.env,
+          setCharm = env.setCharm;
+      setCharm.call(env, props.model.id, upgradeTo, false);
+    },
+
+    render: function() {
+      var items = [],
+          availableVersions = this.props.model.available_versions;
+      if (availableVersions) {
+        availableVersions.forEach(function(version) {
+          items.push(<ns.ChangeVersionItem version={version} upgradeService={this._upgradeService} />)
```

Comment:
Heh yeah it was also tough for me to accept at first but then I had a realization. That JSX isn't HTML it just looks similar, it's actually a DSL for React. This can just as well be written in straight JS but it's a lot more typing.